intermediate_source/transformer_building_blocks.py

```diff
@@ -0,0 +1,641 @@
+solutions. We make the observations that
+
+1.  People want to add slight customizations to their transformer layers
+2.  Writing these layers and customizations is not hard
+
+Supporting all transformer variants via a small number of out of the box layers would
+yield too many keyword arguments. This tutorial will describe how to build your
+own performant transformer layers following our recommended best practices.
+The technologies used will be the following
+
+1.   Nested Tensors with the ``torch.jagged`` layout
+2.   ``scaled_dot_product_attention``
+3.   ``torch.compile()``
+4.   ``FlexAttention``
+
+Is this tutorial for me?
+========================
+
+If you are looking for an out-of-the-box implementation of a popular transformer
+architecture, note that there are many open-source libraries that provide them,
+with some examples being HuggingFace transformers and torchtune. Please head
+there instead!
+
+If you are only interested in performant attention score modifications, please
+head to the `FlexAttention blog <https://flexattention.com/blog/>`_ that
+contains a `gym of masks <https://github.com/pytorch-labs/attention-gym>`_ .
+
+If you are wondering about what building blocks the ``torch`` library provides
+for writing your own transformer layers and best practices, you are in the
+right place, please keep reading!
+
+
+Introducing the Building Blocks
+===============================
+First, we will briefly introduce the 4 technologies mentioned in the introduction
+
+* `torch.nested <https://pytorch.org/tutorials/prototype/nestedtensor.html>`_
+
+Nested tensors generalize the shape of regular dense tensors, allowing for
+representation of ragged-sized data. In the context of transformers,
+we can think of nested tensors as a tool for representing variable sequence
+lengths. They eliminate the need for the bug-prone practices of explicit
+padding and masking (think ``key_padding_mask`` in ``nn.MultiHeadAttention``).
+
+*`scaled_dot_product_attention <https://pytorch.org/tutorials/intermediate/scaled_dot_product_attention_tutorial.html>`_
+
+``scaled_dot_product_attention`` is a primitive for
+$\text{softmax}(\frac{QK^T}{\sqrt{E}} + B)V$ that dispatches into either fused
+implementations of the operator or a fallback implementation. It works out of
+the box in eager mode and also integrates seamlessly with compile.
+As of 2.6, it will also offer grouped query attention natively.
+
+* `torch.compile() <https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html>`_
+
+``torch.compile()`` is a compiler introduced in version 2.0 that is able to
+fuse together sequences of ops. Nested tensors with the ``torch.jagged`` layout
+and ``scaled_dot_product_attention`` work seamlessly with compile. In the
+context of transformers, the value add of using compile with nested tensor
+and sdpa is that compile can remove framework overhead ones sees in eager mode
```

Comment:
nit: might be worthwhile to define eager mode in contrast to torch.compile for clarity